### PR TITLE
New Resource - `azuredevops_check_approval`

### PR DESCRIPTION
**Pull request by josh-barker**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Creating a new resource to manage approvals on service connections, etc

Issue Number:

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [x] Yes
- [ ] No

Notably, I had to modify the vendor copy for the [CheckConfiguration struct](https://github.com/josh-barker/terraform-provider-azuredevops/blob/c601a1668e02edaf91cd4bde044c261fa7c5c60b/vendor/github.com/microsoft/azure-devops-go-api/azuredevops/pipelineschecks/models.go#L38-L39) so that we can set the `Timeout` property on the configuration check.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by josh-barker** [comment]
Hey @xuzhang3, I had to modify the vendor copy for the [CheckConfiguration struct](https://github.com/josh-barker/terraform-provider-azuredevops/blob/c601a1668e02edaf91cd4bde044c261fa7c5c60b/vendor/github.com/microsoft/azure-devops-go-api/azuredevops/pipelineschecks/models.go#L38-L39) so that we can set the Timeout property on the configuration check.

How did you want to tackle this?

**Comment by xuzhang3** [comment]
@josh-barker you can change the vendor files if this is a bug of the SDK. A bit like cracking the SDK.

**Comment by josh-barker** [comment]
Hey @xuzhang3 ,
I forked the azure-devops-go-api repo to fix the struct, so that the lint & unit test jobs can run.

See:
- https://github.com/microsoft/terraform-provider-azuredevops/pull/728/commits/f825d24b994bf64c307968b55cb7d96ae2cb489b
- https://github.com/josh-barker/azure-devops-go-api/commit/0e5ecfc002f70bfecd6457c2dfe8bb674e48298e

Are you okay with that approach, or would you prefer to own the forked repository?

File: azuredevops/internal/service/approvalsandchecks/common_check.go [hunk 1]
```diff
@@ -59,33 +55,32 @@ func genBaseCheckResource(f flatFunc, e expandFunc) *schema.Resource {
...
 	taskCheck := pipelineschecks.CheckConfiguration{
-		Type: &taskCheckType,
-		Settings: map[string]interface{}{
-			"definitionRef": definitionRef,
-			"displayName":   d.Get("display_name").(string),
-			"inputs":        inputs,
-		},
+		Type:     checkType,
+		Settings: settings,
+		// Settings: map[string]interface{}{
```
Comment (xuzhang3) [inline azuredevops/internal/service/approvalsandchecks/common_check.go]: unused code can be deleted.

File: azuredevops/internal/service/approvalsandchecks/resource_check_approval.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+			ValidateFunc: validation.IsUUID,
+		},
+	}
+
+	r.Schema["instructions"] = &schema.Schema{
+		Type:     schema.TypeString,
+		Optional: true,
+	}
+
+	r.Schema["requestor_can_approve"] = &schema.Schema{
```
Comment (xuzhang3) [inline azuredevops/internal/service/approvalsandchecks/resource_check_approval.go]:
```suggestion
	r.Schema["requester_can_approve"] = &schema.Schema{
```

File: azuredevops/internal/service/approvalsandchecks/common_check.go [hunk 3]
```diff
@@ -15,10 +15,6 @@ import (
 	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/tfhelper"
 )
 
-var taskCheckType = pipelineschecks.CheckType{
```
Comment (xuzhang3) [inline azuredevops/internal/service/approvalsandchecks/common_check.go:18]: can we refactor these constant values like and then other resource just refer to it.
```go
type approvalAndCheckTypes struct {
	Approval  *pipelineschecks.CheckType
	TaskCheck *pipelineschecks.CheckType
}

var approvalAndCheckType = approvalAndCheckTypes{
	Approval: &pipelineschecks.CheckType{
		Id:   converter.UUID("8c6f20a7-a545-4486-9777-f762fafe0d4d"),
		Name: converter.ToPtr("Approval"),
	},
	TaskCheck: &pipelineschecks.CheckType{
		Id: converter.UUID("fe1de3ee-a436-41b4-bb20-f6eb4cb879a7"),
	},
}
```

File: azuredevops/internal/service/approvalsandchecks/resource_check_approval.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	settings := map[string]interface{}{
+		"instructions":              d.Get("instructions").(string),
+		"minRequiredApprovers":      d.Get("minimum_required_approvers").(int),
+		"requesterCannotBeApprover": !d.Get("requestor_can_approve").(bool),
+		"approvers":                 approvers,
+	}
+
+	return doBaseExpansion(d, checkTypeApproval, settings, converter.ToPtr(d.Get("timeout").(int)))
```
Comment (xuzhang3) [inline azuredevops/internal/service/approvalsandchecks/resource_check_approval.go]: timeout will be 0 by using `d.Get("timeout")` if `timeout` not set, this is the default value for int values in Terraform. You can determine whether this property is configured by :
```go
timeOutRaw := d.GetRawConfig().AsValueMap()["timeout"]
	if !timeOutRaw.IsNull() {
		timeOut = d.Get("timeout").(int)
	}
```

File: go.mod [hunk 5]
```diff
@@ -62,3 +62,5 @@ require (
 	google.golang.org/grpc v1.48.0 // indirect
 	google.golang.org/protobuf v1.28.1 // indirect
 )
+
+replace github.com/microsoft/azure-devops-go-api/azuredevops => github.com/josh-barker/azure-devops-go-api/azuredevops v1.0.0-b5-checkconfig
```
Comment (xuzhang3) [inline go.mod]: Suggest fix the SDK issue like [pipeline check](https://github.com/microsoft/terraform-provider-azuredevops/blob/main/azuredevops/internal/utils/pipelineschecksextras/pipelineschecks_extras.go) does, patch in this project not in another personal SDK.

File: website/docs/r/check_approval.html.markdown [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+* `target_resource_type` - (Required) TODO. Changing this forces a new Approval Check to be created.
+
+---
+
+* `approvers` - (Required) Specifies a list of TODO.
+
+* `instructions` - (Optional) The instructions for the approvers.
+
+* `minimum_required_approvers` - (Optional) The minimum number of approvers.
```
Comment (xuzhang3) [inline website/docs/r/check_approval.html.markdown]: This property only available when the approver is more than 1, which should be documented.

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

**Comment by josh-barker** [comment]
Hi @xuzhang3 , thanks for your feedback.
I've addressed your comments and also upgraded the API to v7 and created a separate set of functions to use the `GenericConfigurationCheck` struct, as that's not generated in `client.go`.

**Comment by xuzhang3** [comment]
@josh-barker can you revert the v7 SDK upgrade? I tend to upgrade all the resource API from v6 to v7 in a separate PR.

**Comment by josh-barker** [comment]
Hi @xuzhang3 , I've updated the code so that it's aligned to the rest of the other checks. They're all using the v5 API.
I moved the extras to a different directory so that the `azdosdkmocks` can access the packages.

File: .vscode/launch.json [hunk 7]
```diff
@@ -6,14 +6,15 @@
...
+            "program": "${fileDirname}",
             "args": [
                 "-test.v",
                 "-test.run",
-                "^${selectedText}$"
+                "^${selectedText}"
             ],
             "env": {
-                "TF_ACC": "1"
+                "TF_ACC": "1",
```
Comment (xuzhang3) [inline .vscode/launch.json]: Is is personal data?

File: azuredevops/internal/service/approvalsandchecks/resource_check_approval.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+		d.Set("minimum_required_approvers", minRequiredApprovers)
+	} else {
+		return fmt.Errorf("minRequiredApprovers not found")
+	}
+
+	if requesterCannotBeApprover, found := settings["requesterCannotBeApprover"]; found {
+		d.Set("requester_can_approve", !requesterCannotBeApprover.(bool))
+	} else {
+		d.Set("requester_can_approve", true)
+	}
```
Comment (xuzhang3) [inline azuredevops/internal/service/approvalsandchecks/resource_check_approval.go]: `requester_can_approve`  defaults to `false`, no need to set to `true` if not returns be service
```suggestion
	if requesterCannotBeApprover, found := settings["requesterCannotBeApprover"]; found {
		d.Set("requester_can_approve", !requesterCannotBeApprover.(bool))
	} 
```

File: azuredevops/internal/service/approvalsandchecks/resource_check_approval.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+		return fmt.Errorf("settings nil")
+	}
+
+	check.Type = approvalAndCheckType.Approval
+
+	settings := check.Settings.(map[string]interface{})
+
+	if instructions, found := settings["instructions"]; found {
+		d.Set("instructions", instructions)
+	} else {
```
Comment (xuzhang3) [inline azuredevops/internal/service/approvalsandchecks/resource_check_approval.go]: no need to raise an error if `instructions` not found as it is an optional property.

File: azuredevops/internal/service/approvalsandchecks/resource_check_approval.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+		return fmt.Errorf("instructions not found")
+	}
+
+	if minRequiredApprovers, found := settings["minRequiredApprovers"]; found {
+		d.Set("minimum_required_approvers", minRequiredApprovers)
+	} else {
+		return fmt.Errorf("minRequiredApprovers not found")
```
Comment (xuzhang3) [inline azuredevops/internal/service/approvalsandchecks/resource_check_approval.go]: no need to raise an error if instructions not found as it is an optional property.

File: website/docs/r/check_approval.html.markdown [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+* `project_id` - (Required) The project ID. Changing this forces a new Approval Check to be created.
+
+* `target_resource_id` - (Required) The ID of the resource being protected by the check. Changing this forces a new Approval Check to be created.
+
+* `target_resource_type` - (Required) The type of resource being protected by the check. Valid values: `endpoint`, `environment`, `queue`, `repository`, `securefile`, `variablegroup`. Changing this forces a new Approval Check to be created.
+
+---
+
+* `approvers` - (Required) Specifies a list of approvers.
```
Comment (xuzhang3) [inline website/docs/r/check_approval.html.markdown]:
```suggestion
* `approvers` - (Required) Specifies a list of approver IDs.

---

```

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

File: .vscode/launch.json (hunk 7 above)
Comment (josh-barker) [reply .vscode/launch.json]: Whoops! Have updated the file.

**Review by xuzhang3 — APPROVED** [review]

**Comment by xuzhang3** [comment]
```log
=== RUN   TestAccCheckApproval_basic
=== PAUSE TestAccCheckApproval_basic
=== RUN   TestAccCheckApproval_complete
=== PAUSE TestAccCheckApproval_complete
=== RUN   TestAccCheckApproval_update
=== PAUSE TestAccCheckApproval_update
=== RUN   TestAccCheckBranchControl_basic
=== PAUSE TestAccCheckBranchControl_basic
=== RUN   TestAccCheckBranchControl_complete
=== PAUSE TestAccCheckBranchControl_complete
=== RUN   TestAccCheckBranchControl_update
=== PAUSE TestAccCheckBranchControl_update
=== RUN   TestAccCheckBusinessHours_basic
=== PAUSE TestAccCheckBusinessHours_basic
=== RUN   TestAccCheckBusinessHours_complete
=== PAUSE TestAccCheckBusinessHours_complete
=== RUN   TestAccCheckBusinessHours_update
=== PAUSE TestAccCheckBusinessHours_update
=== RUN   TestAccCheckEndpoint
=== PAUSE TestAccCheckEndpoint
=== RUN   TestAccCheckEnvironment
=== PAUSE TestAccCheckEnvironment
=== RUN   TestAccCheckQueue
=== PAUSE TestAccCheckQueue
=== RUN   TestAccCheckRepo
=== PAUSE TestAccCheckRepo
=== RUN   TestAccCheckVariableGroup
=== PAUSE TestAccCheckVariableGroup
=== CONT  TestAccCheckApproval_basic
=== CONT  TestAccCheckBusinessHours_complete
=== CONT  TestAccCheckRepo
=== CONT  TestAccCheckEndpoint
=== CONT  TestAccCheckEnvironment
=== CONT  TestAccCheckQueue
=== CONT  TestAccCheckBranchControl_complete
=== CONT  TestAccCheckVariableGroup
--- PASS: TestAccCheckEnvironment (47.03s)
=== CONT  TestAccCheckBusinessHours_basic
=== CONT  TestAccCheckBranchControl_update
--- PASS: TestAccCheckVariableGroup (48.17s)
=== CONT  TestAccCheckApproval_update
--- PASS: TestAccCheckQueue (48.63s)
--- PASS: TestAccCheckRepo (50.64s)
=== CONT  TestAccCheckBranchControl_basic
--- PASS: TestAccCheckEndpoint (69.15s)
=== CONT  TestAccCheckApproval_complete
--- PASS: TestAccCheckBusinessHours_complete (69.49s)
=== CONT  TestAccCheckBusinessHours_update
--- PASS: TestAccCheckBranchControl_complete (69.75s)
--- PASS: TestAccCheckApproval_basic (75.58s)
--- PASS: TestAccCheckBusinessHours_basic (49.08s)
--- PASS: TestAccCheckBranchControl_basic (48.49s)
--- PASS: TestAccCheckBranchControl_update (56.60s)
--- PASS: TestAccCheckApproval_update (70.32s)
--- PASS: TestAccCheckApproval_complete (54.49s)
--- PASS: TestAccCheckBusinessHours_update (56.78s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        127.077s

```

**Pull request merged**